runtime/src/accounts_db.rs

```diff
@@ -2886,59 +2929,152 @@ impl AccountsDB {
                     "Flushing random slot: {}, num_remaining: {}",
                     *rand_slot, num_slots_remaining
                 );
-                self.flush_slot_cache(*rand_slot);
+                self.flush_slot_cache(*rand_slot, None::<&mut fn(&_, &_) -> bool>, false);
+            }
+        }
+    }
+
+    fn flush_rooted_accounts_cache(
+        &self,
+        mut max_flush_root: Option<Slot>,
+        should_clean: Option<(&mut usize, &mut usize)>,
+    ) -> (usize, usize) {
+        if should_clean.is_some() {
+            max_flush_root = self.max_clean_root(max_flush_root);
+        }
+
+        // If there is a long running scan going on, this could prevent any cleaning
+        // past `max_flush_root`.
+        let cached_roots: BTreeSet<Slot> = self.accounts_cache.clear_roots(max_flush_root);
+        let mut written_accounts = HashSet::new();
+
+        let mut num_roots_flushed = 0;
+
+        // If `should_clean` is None, then`should_flush_f` is also None, which will cause
+        // `flush_slot_cache` to flush all accounts to storage without cleaning any accounts.
+        let mut should_flush_f = should_clean.map(|(account_bytes_saved, num_accounts_saved)| {
+            move |account_key: &Pubkey, account: &Account| {
+                // If a later root already wrote this account, no point
+                // in flushing it
+                let should_flush = !written_accounts.contains(account_key);
+                if should_flush {
+                    written_accounts.insert(*account_key);
+                } else {
+                    *account_bytes_saved += account.data.len();
+                    *num_accounts_saved += 1;
+                }
+                should_flush
+            }
+        });
+
+        // Iterate from highest to lowest so that we don't need to flush earlier
+        // outdated updates in earlier roots
+        for root in cached_roots.iter().rev() {
+            if self.flush_slot_cache(*root, should_flush_f.as_mut(), true) {
```

Comment:
nits: if you like,

```rust
@@ -2969,8 +2968,8 @@ impl AccountsDB {
 
         // Iterate from highest to lowest so that we don't need to flush earlier
         // outdated updates in earlier roots
-        for root in cached_roots.iter().rev() {
-            if self.flush_slot_cache(*root, should_flush_f.as_mut(), true) {
+        for &root in cached_roots.iter().rev() {
+            if self.flush_slot_cache(root, should_flush_f.as_mut(), true) {
                 num_roots_flushed += 1;
             }
         }

```

btw, I learned this syntax today. :)

Reply:
ah yup destructuring, changed :)